numba/cuda/cudadrv/driver.py

```diff
@@ -553,33 +555,370 @@ def met_requirement_for_device(device):
+            self._attempt_allocation(allocator)
+        else:
+            allocator()
+
+        finalizer = _pin_finalizer(self, pointer, mapped)
+        ctx = weakref.proxy(self.context)
+
+        if mapped:
+            mem = MappedMemory(ctx, pointer, size, owner=owner,
+                               finalizer=finalizer)
+            self.allocations[mem.handle.value] = mem
+            return mem.own()
+        else:
+            return PinnedMemory(ctx, pointer, size, owner=owner,
+                                finalizer=finalizer)
+
+    def reset(self):
+        """Clears up all host memory (mapped and/or pinned) in the current
+        context.
+
+        EMM Plugins that override this method must call ``super().reset()`` to
+        ensure that host allocations are also cleaned up."""
+        self.allocations.clear()
+        self.deallocations.clear()
+
+    @contextlib.contextmanager
+    def defer_cleanup(self):
+        """Disables cleanup of mapped or pinned host memory in the current
+        context.
+
+        EMM Plugins that override this method must obtain the context manager
+        from this method before yielding to ensure that cleanup of host
+        allocations is also deferred."""
+        with self.deallocations.disable():
+            yield
+
+
+class NumbaCUDAMemoryManager(HostOnlyCUDAMemoryManager):
+    """Internal on-device memory management for Numba. This is implemented using
+    the EMM Plugin interface, but is not part of the public API."""
+
+    def initialize(self):
+        # Set the memory capacity of *deallocations* as the memory manager
+        # becomes active for the first time
+        if self.deallocations.memory_capacity == _SizeNotSet:
+            self.deallocations.memory_capacity = self.get_memory_info().total
+
+    def memalloc(self, size):
+        ptr = drvapi.cu_device_ptr()
+
+        def allocator():
+            driver.cuMemAlloc(byref(ptr), size)
+
+        self._attempt_allocation(allocator)
+
+        finalizer = _alloc_finalizer(self, ptr, size)
+        ctx = weakref.proxy(self.context)
+        mem = AutoFreePointer(ctx, ptr, size, finalizer=finalizer)
+        self.allocations[ptr.value] = mem
+        return mem.own()
+
+    def get_memory_info(self):
+        free = c_size_t()
+        total = c_size_t()
+        driver.cuMemGetInfo(byref(free), byref(total))
+        return MemoryInfo(free=free.value, total=total.value)
+
+    def get_ipc_handle(self, memory):
+        ipchandle = drvapi.cu_ipc_mem_handle()
+        driver.cuIpcGetMemHandle(
+            byref(ipchandle),
+            memory.owner.handle,
+        )
+        source_info = self.context.device.get_device_identity()
+        offset = memory.handle.value - memory.owner.handle.value
+
+        return IpcHandle(memory, ipchandle, memory.size, source_info,
+                         offset=offset)
+
+    @property
+    def interface_version(self):
+        return 1
+
+
+_memory_manager = None
```

Comment:
Will change this and update all references.

Reply:
Having re-thought this - `_MEMORY_MANAGER` implies a constant to me, but it is not constant (it could be set first). Do you still want it uppercasing given that it can change?

Reply:
ah ok, yes, on second thoughts, that it's non-const suggests lower case is fine, let's leave.